File: day13/part2.py
from functools import reduce
from operator import mul

input = open("input.txt").read().splitlines()

buses = dict()
for offset, bus in enumerate(int(b) if b.isdigit() else None for b in input[1].split(',')):
    if bus:
        buses[bus] = offset

def crt(n, a):
    p = reduce(mul, n)
    return sum(ai * pow(p//ni, -1, ni) * (p//ni) for (ni, ai) in zip(n, a)) % p

print("part2:", reduce(mul, buses.keys()) - crt(buses.keys(), buses.values()))

File: day13/part2.jl
function crt(n::Array, a::Array)
    p = prod(n)
    mod(sum(ai * invmod(p ÷ ni, ni) * (p ÷ ni) for (ni, ai) in zip(n, a)), p)
end

function main()
    input = [
        tryparse(Int, i)!=nothing ? parse(Int, i) : nothing
        for i in split(readlines("input.txt")[end], ",")
    ]

    buses = Dict{Int, Int}()
    for (offset, bus) in enumerate(input)
        if bus != nothing
            buses[bus] = offset-1 # julia is not zero-indexed :D
        end
    end

    println("part2: ", prod(keys(buses))
        - crt(collect(keys(buses)), collect(values(buses))))
end

main()
